### PR TITLE
Discrete entropy

### PR DESCRIPTION
**Pull request by atavory**
Add sample discrete-entropy UDF.

Together with #13574, this forms the basis for mutual information classification udfs #13163.

### CODE REVIEW

**Comment by atavory** [comment]
@rschlussel Could you please review this too? Have rebased it on #13574. Man thanks!

**Comment by ajaygeorge** [comment]
@atavory looks like there are some merge conflicts. Can you rebase with master

File: presto-main/hs_err_pid49551.log [hunk 1]
```diff
@@ -0,0 +1,983 @@
...
+#
+# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
+#
+# If you would like to submit a bug report, please visit:
+#   http://bugreport.java.com/bugreport/crash.jsp
+# The crash happened outside the Java Virtual Machine in native code.
+# See problematic frame for where to report the bug.
+#
+
+---------------  T H R E A D  ---------------
```
Comment (shixuan-fan) [inline presto-main/hs_err_pid49551.log]: This file seems to be unintended changes?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyStateFactory.java [hunk 2]
```diff
@@ -104,9 +104,6 @@ public DifferentialEntropyStateStrategy getStrategy()
         @Override
         public long getEstimatedSize()
         {
-            if (strategy == null) {
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyStateFactory.java]: I have not read the whole PR but why do we remove this null check? `strategy` is set in `setStrategy` so this check seems legit.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyStateStrategy.java [hunk 3]
```diff
@@ -25,35 +31,200 @@
 public interface DifferentialEntropyStateStrategy
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/DifferentialEntropyStateStrategy.java]: Skimmed through this class assuming they are just code moves.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/FixedHistogramMleStateStrategy.java [hunk 4]
```diff
@@ -74,6 +75,14 @@ public void add(double sample, double weight)
         histogram.add(sample, weight);
     }
 
+    // Tmp Ami @Override
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/FixedHistogramMleStateStrategy.java]: nit: we could clean it up :D

File: presto-docs/src/main/sphinx/functions/aggregate.rst [hunk 5]
```diff
@@ -564,6 +564,7 @@ To find the `ROC curve <https://en.wikipedia.org/wiki/Receiver_operating_charact
 
     The thresholds are defined as a sequence whose :math:`j`-th entry is the :math:`j`-th smallest threshold.
 
+
```
[the remark this answers is not in the document]
Comment (atavory) [reply presto-docs/src/main/sphinx/functions/aggregate.rst]: It is actually related to this diff - double space between the start of the new section in the rst file.

File: presto-docs/src/main/sphinx/functions/aggregate.rst [hunk 6]
```diff
@@ -572,7 +573,7 @@ That is, for a random variable :math:`x`, they approximate
 
 .. math ::
 
-    H(x) = - \int x \log_2\left(f(x)\right) dx,
+    h(x) = - \int x \log_2\left(f(x)\right) dx,
```
[the remark this answers is not in the document]
Comment (atavory) [reply presto-docs/src/main/sphinx/functions/aggregate.rst]: With discrete entropy (in this diff), it seems better to be precise and use lowercase "h" for differential entropy.

**Comment by shixuan-fan** [comment]
Why is this closed 😂Is it by accident?

**Comment by rschlussel** [comment]
spoke to @atavory this morning.  He accidentally lost his commit while rebasing. He's gonna restore it and put it back up.

**Pull request closed without merging**